File: opencode/plugins/session-tracker.ts
/**
 * Session Tracker OpenCode Plugin
 *
 * Captures comprehensive session analytics when sessions end.
 *
 * **Migration from**: Claude Code session-tracker.py (SessionEnd)
 *
 * **What it tracks**:
 * - Session duration (start to end)
 * - Total tools used
 * - Unique files modified
 * - Commands executed
 * - Summary of activity
 *
 * **Database**: `~/.opencode/analytics/sessions.db`
 *
 * **Queries you can run**:
 * ```sql
 * -- Recent sessions
 * SELECT * FROM sessions ORDER BY timestamp DESC LIMIT 10;
 *
 * -- Average session duration
 * SELECT AVG(duration_seconds) / 60.0 as avg_minutes FROM sessions;
 *
 * -- Most productive sessions (by tool count)
 * SELECT * FROM sessions ORDER BY tool_count DESC LIMIT 5;
 *
 * -- Total files modified across all sessions
 * SELECT SUM(file_count) as total_files FROM sessions;
 * ```
 *
 * **Note**: Duration and stats calculated from tool_calls database
 *
 * **Test this plugin**:
 * 1. Enable plugin in opencode.json
 * 2. Complete a session with various tool calls
 * 3. Check database: `sqlite3 ~/.opencode/analytics/sessions.db`
 * 4. Run analytics queries
 */

import { Plugin } from '@opencode/plugin-api'
import Database from 'better-sqlite3'
import * as os from 'os'
import * as path from 'path'
import * as fs from 'fs'

export const SessionTrackerPlugin: Plugin = async () => {
  // Database paths
  const dbDir = path.join(os.homedir(), '.opencode', 'analytics')
  const sessionsDbPath = path.join(dbDir, 'sessions.db')
  const toolCallsDbPath = path.join(dbDir, 'tool_calls.db')

  // Ensure directory exists
  if (!fs.existsSync(dbDir)) {
    fs.mkdirSync(dbDir, { recursive: true })
  }

  // Initialize sessions database
  const db = new Database(sessionsDbPath)

  // Create table
  db.exec(`
    CREATE TABLE IF NOT EXISTS sessions (
      id TEXT PRIMARY KEY,
      timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
      duration_seconds INTEGER,
      model TEXT,
      summary TEXT,
      tool_count INTEGER DEFAULT 0,
      file_count INTEGER DEFAULT 0,
      command_count INTEGER DEFAULT 0
    )
  `)

  // Generate session ID
  const sessionId = process.env.OPENCODE_SESSION_ID || `session-${Date.now()}`
  const sessionStartTime = Date.now()

  return {
    session: {
      /**
       * Capture session data when session ends
       */
      end: async () => {
        try {
          // Calculate duration
          const durationSeconds = Math.floor((Date.now() - sessionStartTime) / 1000)

          let toolCount = 0
          let fileCount = 0
          let commandCount = 0

          // Try to get statistics from tool_calls database
          try {
            if (fs.existsSync(toolCallsDbPath)) {
              const toolDb = new Database(toolCallsDbPath, { readonly: true })

              // Count total tools
              const toolResult = toolDb.prepare(`
                SELECT COUNT(*) as count FROM tool_calls WHERE session_id = ?
              `).get(sessionId) as { count: number }
              toolCount = toolResult.count

              // Count unique files
              const fileResult = toolDb.prepare(`
                SELECT COUNT(DISTINCT file_path) as count FROM tool_calls
                WHERE session_id = ? AND file_path IS NOT NULL
              `).get(sessionId) as { count: number }
              fileCount = fileResult.count

              // Count bash commands
              const commandResult = toolDb.prepare(`
                SELECT COUNT(*) as count FROM tool_calls
                WHERE session_id = ? AND tool_type = 'Bash'
              `).get(sessionId) as { count: number }
              commandCount = commandResult.count

              toolDb.close()
            }
          } catch (error) {
            console.warn('[session-tracker] Could not get tool stats:', error.message)
          }

          // Create summary
          const summary = `Session completed: ${toolCount} tools, ${fileCount} files, ${commandCount} commands`

          // Insert session record
          const insertStmt = db.prepare(`
            INSERT OR REPLACE INTO sessions (
              id, model, summary, tool_count, file_count, command_count, duration_seconds
            ) VALUES (?, ?, ?, ?, ?, ?, ?)
          `)

          insertStmt.run(
            sessionId,
            process.env.OPENCODE_MODEL || 'unknown',
            summary,
            toolCount,
            fileCount,
            commandCount,
            durationSeconds
          )

          console.log(`[session-tracker] Session data saved: ${summary}`)
          console.log(`[session-tracker] Duration: ${durationSeconds}s`)

          db.close()
        } catch (error) {
          console.error('[session-tracker] Failed to save session data:', error.message)
        }
      }
    }
  }
}

export default SessionTrackerPlugin

File: claude-code/hooks/session-tracker.py
#!/usr/bin/env python3
# CLAUDE_HOOK_EVENT: SessionEnd
"""
Session End Data Capture - Store comprehensive session analytics

Captures:
- Session duration and timestamp
- Summary of session activity
- Model used
- Session outcome

Creates a historical record of all coding sessions for analytics.
"""

import json
import sys
import sqlite3
import os
from datetime import datetime
from pathlib import Path

# Database paths
SESSIONS_DB = os.path.expanduser("~/.claude/analytics/sessions.db")
TOOL_CALLS_DB = os.path.expanduser("~/.claude/analytics/tool_calls.db")

# Ensure directory exists
os.makedirs(os.path.dirname(SESSIONS_DB), exist_ok=True)

# Read hook input
input_data = json.load(sys.stdin)

# Extract session data
session_id = os.getenv("CLAUDE_SESSION_ID", "unknown")
model = os.getenv("CLAUDE_MODEL", "unknown")

# Initialize database
conn = sqlite3.connect(SESSIONS_DB)
cursor = conn.cursor()

# Create table if not exists
cursor.execute("""
    CREATE TABLE IF NOT EXISTS sessions (
        id TEXT PRIMARY KEY,
        timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
        duration_seconds INTEGER,
        model TEXT,
        summary TEXT,
        tool_count INTEGER DEFAULT 0,
        file_count INTEGER DEFAULT 0,
        command_count INTEGER DEFAULT 0
    )
""")

# Try to get statistics from tool_calls database
tool_count = 0
file_count = 0
command_count = 0
duration_seconds = 0

try:
    tool_conn = sqlite3.connect(TOOL_CALLS_DB)
    tool_cursor = tool_conn.cursor()

    # Count total tools used
    tool_cursor.execute("""
        SELECT COUNT(*) FROM tool_calls WHERE session_id = ?
    """, (session_id,))
    tool_count = tool_cursor.fetchone()[0]

    # Count unique files modified
    tool_cursor.execute("""
        SELECT COUNT(DISTINCT file_path) FROM tool_calls
        WHERE session_id = ? AND file_path IS NOT NULL
    """, (session_id,))
    file_count = tool_cursor.fetchone()[0]

    # Count bash commands
    tool_cursor.execute("""
        SELECT COUNT(*) FROM tool_calls
        WHERE session_id = ? AND tool_type = 'Bash'
    """, (session_id,))
    command_count = tool_cursor.fetchone()[0]

    # Calculate session duration
    tool_cursor.execute("""
        SELECT
            CAST((julianday(MAX(timestamp)) - julianday(MIN(timestamp))) * 24 * 60 * 60 AS INTEGER)
        FROM tool_calls WHERE session_id = ?
    """, (session_id,))
    result = tool_cursor.fetchone()
    duration_seconds = result[0] if result[0] else 0

    tool_conn.close()
except Exception as e:
    print(f"[session-tracker] Warning: Could not get tool stats: {e}", file=sys.stderr)

# Create simple summary
summary = f"Session completed: {tool_count} tools, {file_count} files, {command_count} commands"

# Insert session record
cursor.execute("""
    INSERT OR REPLACE INTO sessions (
        id, model, summary, tool_count, file_count, command_count, duration_seconds
    ) VALUES (?, ?, ?, ?, ?, ?, ?)
""", (
    session_id,
    model,
    summary,
    tool_count,
    file_count,
    command_count,
    duration_seconds
))

conn.commit()
conn.close()

print(f"[session-tracker] Session data saved: {summary}", file=sys.stderr)
print(f"[session-tracker] Duration: {duration_seconds}s", file=sys.stderr)

sys.exit(0)
